### ISSUE TITLE
Can not delete directory from zip archive bug

### ISSUE THREAD

**Issue by VladimirSmall** [icsharpcode/SharpZipLib#82]
ICSharpCode.SharpZipLib version - 0.86.0.518.

Code to demonstrate the bug:

``` csharp
namespace SharpZipLibDeleteFolderBug
{
    using ICSharpCode.SharpZipLib.Zip;

    using NUnit.Framework;

    [TestFixture]
    public class SharpZipLibDeleteFolderBugDemo
    {
        [Test]
        public void Bug()
        {
            var zipFile = ZipFile.Create("SomeFile.zip");
            CreateDirectory(zipFile, "SomeDirectory");

            DeleteDirectory(zipFile, "SomeDirectory");
        }

        private void DeleteDirectory(ZipFile zipFile, string directoryName)
        {
            zipFile.BeginUpdate();

            var zipEntry = zipFile.GetEntry(directoryName + "/");
            Assert.That(zipEntry, Is.Not.Null); // passed - directory was successfully create

            zipFile.Delete(zipEntry); // !!! throws ICSharpCode.SharpZipLib.Zip.ZipException : Cannot find entry to delete

            zipFile.CommitUpdate();            
        }

        private static void CreateDirectory(ZipFile zipFile, string directoryName)
        {
            zipFile.BeginUpdate();
            zipFile.AddDirectory(directoryName);
            zipFile.CommitUpdate();
        }
    }
}
```

The buggy method is ZipFile.FindExistingUpdate:

``` csharp
        int FindExistingUpdate(ZipEntry entry)
        {
            int result = -1;
            string convertedName = GetTransformedFileName(entry.Name); // !!! BUG is here

            if (updateIndex_.ContainsKey(convertedName)) {
                result = (int)updateIndex_[convertedName];
            }

                        ...
            return result;
        }
```

Method FindExistingUpdate ignores the fact that entry can be not only file entry but also a directory entry. Right implementation should be:

``` csharp
        int FindExistingUpdate(ZipEntry entry)
        {
            int result = -1;
            string convertedName;
            if (entry.IsDirectory)
                convertedName = GetTransformedDirectoryName(entry.Name);
            else 
                convertedName = GetTransformedFileName(entry.Name);

            if (updateIndex_.ContainsKey(convertedName)) {
                result = (int)updateIndex_[convertedName];
            }

                        ...
            return result;
        }
```

Method ZipFile.Delete(string fileName) also has the same bug.

**Comment by Numpsy** [comment]
Fixed by #295 ? (and maybe the same as #48)

**Comment by Numpsy** [comment]
are this and/or #48 fixed now, or is #402 still part of the same issue?

**Comment by piksel** [comment]
I'm not sure. This is basically the main thing left before 1.4, solving the name-mapping (with a defined behaviour) och updating the currently failing tests.
